xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiServletResponse.java

```diff
@@ -62,34 +65,25 @@ public HttpServletResponse getHttpServletResponse()
     @Override
     public void sendRedirect(String redirect) throws IOException
     {
-        if (StringUtils.isBlank(redirect)) {
-            // Nowhere to go to
-            return;
-        }
-        if (StringUtils.containsAny(redirect, '\r', '\n')) {
-            LOGGER.warn("Possible HTTP Response Splitting attack, attempting to redirect to [{}]", redirect);
-            return;
-        }
-
-        if (StringUtils.startsWith(redirect, "//")) {
-            LOGGER.warn("Possible phishing attack, attempting to redirect to [{}]. If this request is legitimate, "
-                + "use an actual absolute URL and pay attention to configure properly url.trustedDomains in "
-                + "xwiki.properties", redirect);
-            return;
-        }
-
-        // check for trusted domains, only if the given location is an absolute URL.
-        if (ABSOLUTE_URL_PATTERN.matcher(redirect).matches()) {
-            if (!getURLSecurityManager().isDomainTrusted(new URL(redirect))) {
-                LOGGER.warn(
-                    "Possible phishing attack, attempting to redirect to [{}], this request has been blocked. "
-                        + "If the request was legitimate, add the domain related to this request in the list "
-                        + "of trusted domains in the configuration: it can be configured in xwiki.properties in "
-                        + "url.trustedDomains.", redirect);
-                return;
+        if (!StringUtils.isBlank(redirect)) {
+            try {
+                URI uri = new URI(redirect);
+                if (!getURLSecurityManager().isURITrusted(uri)) {
+                    LOGGER.warn(
+                        "Possible phishing attack, attempting to redirect to [{}], this request has been blocked. "
+                            + "If the request was legitimate, please check the URL security configuration. You "
+                            + "might need to add the domain related to this request in the list of trusted domains in "
+                            + "the configuration: it can be configured in xwiki.properties in url.trustedDomains.",
+                        redirect);
+                } else {
+                    this.response.sendRedirect(redirect);
```

Comment:
I'm wondering if the redirect should use the `uri.toString()` to ensure that the parser didn't do any normalization that might be done differently by a browser but then again I'm not sure it's worth it and if there is really so much magic in `URI`.

Reply:
At least for now I wouldn't change that as it would introduce possibly an unwanted behaviour in redirect. Here we're sure the user is redirected to the indicated string, and the security component acts only as an oracle. We could in the future decide to transform the indicated redirect to another one, but the API should be clear it performs a transformation.